File: CopyPasteDemo/CopyPasteDemo/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace CopyPasteDemo
{
    public partial class frmContractList : Form
    {
        public frmContractList()
        {
            InitializeComponent();
        }

        // Initialize ListView
        private void InitializeListView()
        {
            // Set the view to show details.
            lstContractList.View = View.Details;

            // Allow the user to edit item text.
            lstContractList.LabelEdit = true;

            // Allow the user to rearrange columns.
            lstContractList.AllowColumnReorder = true;

            // Select the item and subitems when selection is made.
            lstContractList.FullRowSelect = true;

            // Display grid lines.
            lstContractList.GridLines = true;

            //lstContractList.

            // Sort the items in the list in ascending order.
            lstContractList.Sorting = SortOrder.Ascending;

            // Attach Subitems to the ListView
            lstContractList.Columns.Add("Contract", 120, HorizontalAlignment.Left);
            lstContractList.Columns.Add("Logo", 40, HorizontalAlignment.Left);
            lstContractList.Columns.Add("Status", 185, HorizontalAlignment.Left);

        }

        private void Form1_Load(object sender, EventArgs e)
        {
            InitializeListView();
            // Set the view to show details.
            //lstContractList.View = View.Details;
            //lstContractList.Columns.Add("Contract", 170, HorizontalAlignment.Left);
            //lstContractList.Columns.Add("Stautus", 185, HorizontalAlignment.Left);
            AddItem(lstContractList, "102-01", Properties.Resources.Ready, "Completed");
            AddItem(lstContractList, "102", Properties.Resources.Error, "Error");
            AddItem(lstContractList, "102-02", Properties.Resources.Warning, "Invalid");
        }


        // Make a server status item.
        private void AddItem(ListView lvw, string server, Image logo, string desc)//Color status,
        {
            // Make the item.
            ListViewItem item = new ListViewItem(server);

            // Save the ServeStatus item in the Tag property.
            Data server_status = new Data(server, logo, desc);// status,
            item.Tag = server_status;
            item.SubItems[0].Name = "Contract";

            // Add subitems so they can draw.
            item.SubItems.Add("Logo");
            item.SubItems.Add("Status");
            //item.SubItems.Add("Desc");

            // Add the item to the ListView.
            lvw.Items.Add(item);
        }

        private void lstContractList_MouseDown(object sender, MouseEventArgs e)
        {

        }

        private void lstContractList_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.Control && e.KeyCode == Keys.C)
            {
                // add your processing code, this is an example 
                MessageBox.Show("You have pressed ctrl-c!");
            }
            if (e.Control && e.KeyCode == Keys.V)
            {
                // add your processing code, this is an example 
                MessageBox.Show("You have pressed ctrl-v!");
                PasteClipboard(lstContractList);
                //BindListViewItems();
            }
        }

        private void BindListViewItems()
        {
            //lstContractList.Items.Clear();
            List<Data> item = new List<Data>()
            {
                new Data { Contract = "1", Status = "Completed" }, new Data { Contract = "2", Status = "Closed" }, new Data { Contract = "3", Status = "Pre-close" },
                };
            LoadList(item);
        }

        private void PasteClipboard(ListView myListView)
        {
            DataObject dObject = (DataObject)Clipboard.GetDataObject();
            if (dObject.GetDataPresent(DataFormats.Text))
            {
                string[] pastedRows = Regex.Split(dObject.GetData(DataFormats.Text).ToString().TrimEnd("\r\n".ToCharArray()), "\r\n");
                List<Data> contractList = new List<Data>();
                foreach (string pastedRow in pastedRows)
                {
                    string[] pastedRowCells = pastedRow.Split(new char[] { '\t' });

                    if (pastedRowCells.Length > 0)
                        contractList.Add(new Data { Contract = pastedRowCells[0] }); //, Status = pastedRowCells[1]

                }

                //list of status
                List<Data> itemStatus = new List<Data>()
                {
                    new Data { Contract = "1", Logo=Properties.Resources.Error, Status = "Error" }, new Data { Contract = "4", Logo=Properties.Resources.Error, Status = "Closed" }, new Data { Contract = "2", Logo=Properties.Resources.Ready, Status = "Completed" },
                    new Data { Contract = "3", Logo=Properties.Resources.Warning, Status = "INVALID" },
                };

                foreach (var i in itemStatus)
                {
                    var itemToChange = contractList.First(c => c.Contract == i.Contract).Status = i.Status;
                    var logo = contractList.First(c => c.Contract == i.Contract).Logo = i.Logo;
                }

                LoadList(contractList);
            }
        }
       
        private void LoadList(List<Data> item)
        {

            ListViewItem lvi;
            //Add each Row as a ListViewItem 
            foreach (Data data in item)
            {
                lvi = new ListViewItem(data.Contract);
                Data dataItem = new Data(data.Contract, data.Logo, data.Status);
                lvi.Tag = data;
                lvi.SubItems[0].Name = "Contract";
                // Add subitems so they can draw.
                lvi.SubItems.Add("Contract");
                lvi.SubItems.Add("Status");
                //lvi.SubItems.Add(data.Logo);
                //lvi.SubItems.Add(data.Status);
                //lvi.Tag = data;
                lstContractList.Items.Add(lvi);
            }
        }

        private void lstContractList_KeyUp(object sender, KeyEventArgs e)
        {

        }

        private void btnAdd_Click(object sender, EventArgs e)
        {
            lstContractList.Items.Add(new ListViewItem(txtContract.Text));
            txtContract.Clear();
        }

        private void btnRemove_Click(object sender, EventArgs e)
        {
            if (lstContractList.SelectedItems.Count > 0)
            {
                var confirmation = MessageBox.Show("Are you sure ?", "Confirmation", MessageBoxButtons.YesNo, MessageBoxIcon.Question);
                if (confirmation == DialogResult.Yes)
                {
                    foreach (ListViewItem item in lstContractList.SelectedItems)
                    {
                        lstContractList.Items.Remove(item);
                    }
                    //for (int i = lstContractList.SelectedItems.Count - 1; i >= 0; i--)
                    //{
                    //    lstContractList.Items[lstContractList.SelectedItems[i].Index].Remove();
                    //}
                }
            }
            else
                MessageBox.Show("Contract not selected");
        }

        private void btnRemoveAll_Click(object sender, EventArgs e)
        {
            lstContractList.Items.Clear();
        }

        private void btnPreClose_Click(object sender, EventArgs e)
        {

        }

        private void lstContractList_DrawColumnHeader(object sender, DrawListViewColumnHeaderEventArgs e)
        {
            using (StringFormat string_format = new StringFormat())
            {
                string_format.Alignment = StringAlignment.Center;
                string_format.LineAlignment = StringAlignment.Center;

                string text = lstContractList.Columns[e.ColumnIndex].Text;
                switch (e.ColumnIndex)
                {
                    case 0:
                        e.Graphics.DrawString(text, lstContractList.Font, Brushes.Black, e.Bounds);
                        break;
                    case 1:
                        e.Graphics.DrawString(text, lstContractList.Font, Brushes.Black, e.Bounds);
                        break;
                    case 2:
                        e.Graphics.DrawString(text, lstContractList.Font, Brushes.Black, e.Bounds);
                        break;
                    //case 3:
                    //    e.Graphics.DrawString(text, lstContractList.Font, Brushes.Black, e.Bounds);
                    //    break;
                }
            }
        }

        private void lstContractList_DrawItem(object sender, DrawListViewItemEventArgs e)
        {           
            // Get the ListView item and the ServerStatus object.
            ListViewItem item = e.Item;
            Data server_status = item.Tag as Data;

            // Clear.
            e.DrawBackground();


            // Draw the image.
            e.Graphics.DrawImage(server_status.Logo, 0, 0);

            // Draw the focus rectangle if appropriate.
            e.Graphics.ResetTransform();
            e.DrawFocusRectangle();
        }

        private void lstContractList_DrawSubItem(object sender, DrawListViewSubItemEventArgs e)
        {
            // Get the ListView item and the ServerStatus object.
            ListViewItem item = e.Item;
            Data server_status = item.Tag as Data;

            // Draw.
            switch (e.ColumnIndex)
            {
                case 0:
                    // Draw the server's name.
                    e.Graphics.DrawString(server_status.Contract,
                        lstContractList.Font, Brushes.Black, e.Bounds);
                    break;
                case 1:
                    // Draw the server's logo.
                    float scale = e.Bounds.Height / (float)server_status.Logo.Height;
                    e.Graphics.ScaleTransform(scale, scale);
                    e.Graphics.TranslateTransform(
                        e.Bounds.Left,
                        e.Bounds.Top + (e.Bounds.Height - server_status.Logo.Height * scale) / 2,
                        System.Drawing.Drawing2D.MatrixOrder.Append);
                    e.Graphics.DrawImage(server_status.Logo, 0, 0);
                    break;
                case 2:
                    // Draw the server's name.
                    e.Graphics.DrawString(server_status.Status,
                        lstContractList.Font, Brushes.Black, e.Bounds);
                    break;
            }

            // Draw the focus rectangle if appropriate.
            e.Graphics.ResetTransform();
            ListView lvw = e.Item.ListView;
            if (lvw.FullRowSelect)
            {
                e.DrawFocusRectangle(e.Item.Bounds);
            }
            else if (e.SubItem.Name == "Server")
            {
                e.DrawFocusRectangle(e.Bounds);
            }
        }
    }

    class Data
    {
        public string Contract { get; set; }
        public Image Logo;
        public string Status { get; set; }
        public Data()
        { }

        public Data(string contract, Image logo, string status)
        {
            Contract = contract;
            Logo = logo;
            Status = status;
        }
    }
}

File: CopyPasteDemo/VBCopyPasteDemo/Form1.vb
Imports System.Data.SqlClient
Imports System.Text
Imports System.Text.RegularExpressions

Public Class frmContract

    Dim contractFilteredList As New List(Of ContractStatus)()

    Private Sub btnRemove_Click(sender As Object, e As EventArgs) Handles btnRemove.Click
        If lstContractList.SelectedItems.Count > 0 Then
            'Dim confirmation = MessageBox.Show("Are you sure ?", "Confirmation", MessageBoxButtons.YesNo, MessageBoxIcon.Question)
            'If confirmation = DialogResult.Yes Then
            For Each item As ListViewItem In lstContractList.SelectedItems
                lstContractList.Items.Remove(item)
                'for (int i = lstContractList.SelectedItems.Count - 1; i >= 0; i--)
                '{
                '    lstContractList.Items[lstContractList.SelectedItems[i].Index].Remove();
                '}
            Next
            'End If
            'Else
            '    MessageBox.Show("Contract not selected")
        End If
    End Sub

    Private Sub btnRemoveAll_Click(sender As Object, e As EventArgs) Handles btnRemoveAll.Click
        lstContractList.Items.Clear()
    End Sub

    Private Sub btnPreClose_Click(sender As Object, e As EventArgs) Handles btnPreClose.Click
        Dim count As Integer = contractFilteredList.FindAll(Function(x) x.Status = "Completed").Count
        If count > 0 Then
            MsgBox("There there are " + count.ToString() + " number of contracts that cannot be pre-closed")
        End If
    End Sub

    Private Sub btnAdd_Click(sender As Object, e As EventArgs) Handles btnAdd.Click
        'get the status and bind
        Dim contractDataList As New List(Of ContractStatus)()
        GetContractList(txtContract.Text.Trim(), contractDataList)
        LoadList(contractDataList)
        contractFilteredList.AddRange(contractDataList)
        'lstContractList.Items.Add(New ListViewItem(txtContract.Text))
        txtContract.Clear()
    End Sub

    Private Sub frmContract_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        InitializeListView()
        AddContextMenu()
        'AddContextMenuAndItems()
    End Sub

    Public Sub AddContextMenu()
        Dim menuStrip As New ContextMenuStrip()
        Dim menuItem As New ToolStripMenuItem("Exit")
        menuItem.Name = "Exit"
        'menuItem.ShortcutKeys = Keys.Alt|Keys.
        ' Associate an event handler with an event.
        AddHandler menuItem.Click, AddressOf Me.menuItem_Click
        menuStrip.Items.Add(menuItem)
        Me.ContextMenuStrip = menuStrip
        Dim menuItem1 As New ToolStripMenuItem("Remove")
        menuItem1.Name = "Remove"
        AddHandler menuItem1.Click, AddressOf Me.menuItem_Click
        menuStrip.Items.Add(menuItem1)
        Me.ContextMenuStrip = menuStrip
        Dim menuItem2 As New ToolStripMenuItem("Paste")
        menuItem2.Name = "Paste"
        AddHandler menuItem2.Click, AddressOf Me.menuItem_Click
        menuStrip.Items.Add(menuItem2)
        Me.ContextMenuStrip = menuStrip
    End Sub

    Protected Sub menuItem_Click(sender As Object, e As EventArgs)
        Dim menuItem As ToolStripItem = DirectCast(sender, ToolStripItem)
        If menuItem.Name = "Exit" Then
            Application.[Exit]()
        End If
    End Sub

    Public Sub AddContextMenuAndItems()
        Dim mnuContextMenu As New ContextMenu()
        Me.ContextMenu = mnuContextMenu
        Dim mnuItemRemove As New MenuItem()
        mnuItemRemove.Text = "&Remove"
        mnuContextMenu.MenuItems.Add(mnuItemRemove)

        Dim mnuItemRemoveAll As New MenuItem()
        mnuItemRemoveAll.Text = "Remove &All"
        mnuContextMenu.MenuItems.Add(mnuItemRemoveAll)

        Dim mnuItemRemoveInValid As New MenuItem()
        mnuItemRemoveInValid.Text = "Remove &Invalid"
        mnuContextMenu.MenuItems.Add(mnuItemRemoveInValid)

        Dim mnuItemPaste As New MenuItem()
        mnuItemPaste.Text = "&Paste"
        mnuContextMenu.MenuItems.Add(mnuItemPaste)

        lstContractList.ContextMenu = mnuContextMenu

    End Sub

    Private Sub InitializeListView()

        ' Set the view to show details.
        lstContractList.View = View.Details

        ' Allow the user to edit item text.
        lstContractList.LabelEdit = True

        ' Allow the user to rearrange columns.
        lstContractList.AllowColumnReorder = True

        ' Select the item And subitems when selection Is made.
        lstContractList.FullRowSelect = True

        ' Display grid lines.
        lstContractList.GridLines = True

        'lstContractList.

        ' Sort the items in the list in ascending order.
        lstContractList.Sorting = SortOrder.Ascending

        ' Attach Subitems to the ListView
        lstContractList.Columns.Add("Contract", 170, HorizontalAlignment.Left)
        lstContractList.Columns.Add("Status", 185, HorizontalAlignment.Left)

    End Sub

    Private Sub lstContractList_KeyDown(sender As Object, e As KeyEventArgs) Handles lstContractList.KeyDown
        'ctrl-c
        If e.Control And e.KeyCode = Keys.C Then
            ' add your processing code, this is an example 
        End If
        'ctrl-v
        If e.Control And e.KeyCode = Keys.V Then
            PasteClipboard(lstContractList)
        End If
    End Sub


    Private Sub PasteClipboard(myListView As ListView)
        'Get the copied item to be paste
        Dim dObject As DataObject = DirectCast(Clipboard.GetDataObject(), DataObject)
        If dObject.GetDataPresent(DataFormats.Text) Then
            Dim pastedRows As String() = Regex.Split(dObject.GetData(DataFormats.Text.Trim).ToString().TrimEnd(vbCrLf.ToCharArray()), vbCrLf)
            Dim contractList As New List(Of ContractStatus)()
            Dim contractID As String = String.Empty
            Dim contractDataList As New List(Of ContractStatus)()
            For Each pastedRow As String In pastedRows
                Dim pastedRowCells As String() = pastedRow.Split(New Char() {ControlChars.Tab})

                If pastedRowCells.Length > 0 Then
                    contractList.Add(New ContractStatus() With {.Contract = pastedRowCells(0)})   ', Status = pastedRowCells[1]
                    contractID += pastedRowCells(0) + ","
                End If
            Next

            contractID = contractID.Remove(contractID.Length - 1)
            'remove the last comma from the stringbuilder
            'sbContractID.Length = sbContractID.Length - 1
            'sbContractID.Remove(sbContractID.Length - 1, 1)
            'sbContractID.ToString().TrimEnd(',')
            GetContractList(contractID, contractDataList)


            '    'list of status should get from DB by passing the contract
            '    Dim itemStatus As New List(Of Data)() From {
            '    New Data() With {
            '        .Contract = "1",
            '        .Status = "Completed"
            '    },
            '    New Data() With {
            '        .Contract = "4",
            '        .Status = "Closed"
            '    },
            '    New Data() With {
            '        .Contract = "2",
            '        .Status = "Pre-close"
            '    },
            '    New Data() With {
            '        .Contract = "3",
            '        .Status = "INVALID"
            '    }
            '}

            '    For Each i As Data In itemStatus
            '        contractList.First(Function(c) c.Contract = i.Contract).Status = i.Status
            '    Next

            'LoadList(contractList)
            LoadList(contractDataList)
            contractFilteredList.AddRange(contractDataList)
        End If
    End Sub

    Private Shared Sub GetContractList(contractID As String, contractDataList As List(Of ContractStatus))
        Dim connetionString As String
        Dim connection As SqlConnection
        Dim command As New SqlCommand
        Dim param As SqlParameter
        'Dim sql As String
        connetionString = "Data Source=SAROJSAHU-PC;Initial Catalog=FSL2UK02;User ID=sa;Password=Exch3qu3r#1" 'Windows Auth-"Server= localhost; Database= employeedetails;Integrated Security=SSPI;"
        'sql = "SELECT CP.[ContractID] As Contract,CS.ContractStatus AS Status FROM [dbo].[tblContract_PROTECTED] CP INNER JOIN [dbo].tblContractStatus CS ON CP.ContractStatusID = CS.[ContractStatusID] 
        '            WHERE [ContractID] IN ( SELECT t.c.value('.', 'VARCHAR(20)')
        '	            FROM (SELECT x = CAST('<t>' + REPLACE('" + contractID + "', ',', '</t><t>') + '</t>' AS XML)
        '                                          ) a
        '                                  CROSS APPLY x.nodes('/t') t(c))"
        'sql = "SELECT * INTO #TempContract FROM dbo.SplitString('" + contractID + "',',') 
        '    SELECT t.Item [Contract], ISNULL((SELECT cs.ContractStatus FROM tblContractStatus AS cs WHERE cs.ContractStatusID = c.ContractStatusID),'INVALID') [Status]
        '    FROM [tblContract_PROTECTED] AS c
        '    RIGHT JOIN #TempContract AS t ON c.ContractID=t.Item 
        '    DROP TABLE #TempContract"
        'Setup SQL Command
        connection = New SqlConnection(connetionString)
        Try
            connection.Open()
            'command = New SqlCommand(sql, connection)

            ' Create the command with the sproc name and add the parameter required
            command.Connection = connection
            command.CommandType = CommandType.StoredProcedure
            command.CommandText = "fsspGetContractStatusByContractID"
            param = New SqlParameter("@ContractID", contractID)
            param.Direction = ParameterDirection.Input
            param.DbType = DbType.String
            command.Parameters.Add(param)

            'command to create an SqlDataReader on the result of the sproc'
            Dim sqlReader As SqlDataReader = command.ExecuteReader()
            While sqlReader.Read()
                ' Get the first and second field frm the reader'
                contractDataList.Add(New ContractStatus() With {.Contract = sqlReader(0).ToString(), .Status = sqlReader(1).ToString()})
            End While
            sqlReader.Close()
            command.Dispose()
            connection.Close()
        Catch ex As Exception
            MsgBox(ex.Message)
        End Try
    End Sub

    Private Sub LoadList(item As List(Of ContractStatus))

        'Create two ImageList objects.
        Dim imageListSmall As New ImageList()
        ' Initialize the ImageList objects with bitmaps.
        imageListSmall.Images.Add(Bitmap.FromFile("C:\Working\Exchequer Bespoke\Bespoke\White House Products Ltd\Advanced.Bespoke.Whitehouse\Icons\Error.png"))
        imageListSmall.Images.Add(Bitmap.FromFile("C:\Working\Exchequer Bespoke\Bespoke\White House Products Ltd\Advanced.Bespoke.Whitehouse\Icons\Ready.png"))
        'Assign the ImageList objects to the ListView.
        lstContractList.SmallImageList = imageListSmall

        Dim lvi As ListViewItem
        'Add each Row as a ListViewItem 
        For Each data As ContractStatus In item
            lvi = New ListViewItem(data.Contract)
            lvi.SubItems.Add(data.Status)
            lvi.Tag = data

            ' Sets the first list item to display the 4th image.

            lstContractList.Items.Add(lvi)
            If data.Status = "INVALID" Then
                lstContractList.Items(lstContractList.Items.Count - 1).ImageIndex = 0
            Else
                lstContractList.Items(lstContractList.Items.Count - 1).ImageIndex = 1

            End If
            ' Add the ListView to the control collection.
            'Me.Controls.Add(lstContractList)
        Next

        'lstContractList

    End Sub



    Private Sub lstContractList_AfterLabelEdit(sender As Object, e As LabelEditEventArgs) Handles lstContractList.AfterLabelEdit
        ' Determine if label is changed by checking to see if it is equal to Nothing.
        If e.Label Is Nothing Then
            Return
        End If
        ' ASCIIEncoding is used to determine if a number character has been entered.
        Dim AE As New ASCIIEncoding()
        ' Convert the new label to a character array.
        Dim temp As Char() = e.Label.ToCharArray()

        'lstContractList.FindItemWithText(lstContractList.)


        ' Check each character in the new label to determine if it is a number.
        Dim x As Integer
        For x = 0 To temp.Length - 1
            ' Encode the character from the character array to its ASCII code.
            Dim bc As Byte() = AE.GetBytes(temp(x).ToString())

            ' Determine if the ASCII code is within the valid range of numerical values.
            If bc(0) > 47 And bc(0) < 58 Then
                ' Cancel the event and return the lable to its original state.
                e.CancelEdit = True
                ' Display a MessageBox alerting the user that numbers are not allowed.
                MessageBox.Show("The text for the item cannot contain numerical values.")
                ' Break out of the loop and exit.
                Return
            End If
        Next x
    End Sub
End Class


Class ContractStatus
    Public Property Contract() As String
        Get
            Return m_Contract
        End Get
        Set
            m_Contract = Value
        End Set
    End Property
    Private m_Contract As String
    Public Property Status() As String
        Get
            Return m_Status
        End Get
        Set
            m_Status = Value
        End Set
    End Property
    Private m_Status As String
End Class
